[로또] 올레 미션 제출합니다.

jinwookh opened the pull request:
푸시 이후에도 간간히 수정할 예정입니다!

kimjungmin-developer committed:
범위 밖 숫자입력 예외처리 기능 구현.

kimjungmin-developer committed:
로또 입력(중복, 부적절한 숫자 개수) 예외 처리 기능 구현

kimjungmin-developer committed:
구입 금액에 대한 예외 처리 기능 구현.

kimjungmin-developer committed:
Generate lottos

kimjungmin-developer committed:
get prize rank by counting lotto number match

kimjungmin-developer committed:
Calculate prize percentage of lottos

kimjungmin-developer committed:
get input of money, count of Lotto, lotto, winning Lotto

kimjungmin-developer committed:
print statistics

jinwookh committed:
refactor code by convention

jinwookh committed:
change type of lotto numbers data, list to map

jinwookh committed:
refactor constant variable

jinwookh committed:
delete function of repetitive asking input(delete recursive method)

jinwookh committed:
change new function to static create function of lotto, money, winninglotto class

jinwookh committed:
calculate profit rate by percentage(100% is no profit, no loss)

jinwookh committed:
code refactored by convention

jihan805 commented on src/main/java/lotto/controller/LottoLauncher.java:
```diff
@@ -0,0 +1,36 @@
+    public static void main(String[] args) {
+        Money money = Money.create(InputView.askMoney());
+        Lottos manualLottos = getManualLottos(money);
+        Lottos automaticLottos = LottoMachine.generateLottos(money.getLottoCount() - manualLottos.size());
```
현재 Main에서 로또 생성 로직 일부를 담당하고 있어요!
로또 생성은 자동/수동 방식이 있어요.
둘다 로또를 생성한다는 공통점이 있는데 인터페이스로 충분히 추출할 수 있지 않을까요?

jinwookh replied:
넵 generate() 인터페이스를 만들어, 자동과 수동 클래스에서 각자 다르게 구현하는 방식을 생각해 보았습니다. 
여기에 대해 두 가지 고민이 있습니다.
1. 수동 방식은 자동 방식과 받는 정보가 다릅니다. 수동 방식에게는 로또 숫자들에 대한 정보를 넘겨 주어야 하고, 자동 방식에게는 생성할 로또 수만 넘겨주면 됩니다. 따라서 수동 방식, 자동 방식은 받는 매개변수의 형식에 있어서 차이가 생기게 됩니다. 매개변수 형식이 다른데 인터페이스로 함께 묶을 수 있는 방법이 있을까요? 
2. 수동 방식에서는 수동 로또 개수만큼 콘솔에 입력받는 부분이 있습니다. 이 부분은 view와 관계되어 있다고 생각되어, 수동 로또 생성하는 로직과 어떻게든 분리하고 싶은데, 방법이 잘 떠오르지 않습니다. ㅠㅠ
위 고민들에 대해 Summer 리뷰어님의 팁을 받을 수 있을까요?

jihan805 replied:
고민 👍 
1. 생성자를 활용해보면 어떨까요?
2. view로 부터 수동 로또 만큼의 리스트를 받도록 구현하고, 수동 로또에서는 이 리스트를 이용해보면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,49 @@
+
+    public static List<LottoNumber> convertToLottoNumbers(List<Integer> numbers) {
+        List<LottoNumber> returnList = new ArrayList<>();
+        for (int i = 0, n = numbers.size(); i < n; i++) {
```
for each를 사용할 수 있을 것 같아요!

jinwookh replied:
넵!

jihan805 commented on src/main/java/lotto/domain/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+
+public class WinningLotto {
+    private Lotto lotto;
```
WinningLotto 클래스 추출 👍
로또는 구매한 로또, 우승번호를 가진 로또 두가지 종류가 있어요!
WinningLotto클래스 내부이긴 하지만, 변수이름 역시 winningLotto와 같이 작성하면 코드를 보면서 조금 더 명확하게 알 수 있지 않을까요?

jinwookh replied:
네 winningLotto라는 변수명이 더 명확한 것 같아요!

jihan805 commented on src/main/java/lotto/domain/money/Money.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public static Money create(int money) {
+        if (money < Lotto.PRICE || money % Lotto.PRICE != 0) {
```
상수 사용!
전체적으로 상수를 활용할 수 있는 부분을 찾아서 적용해보면 어떨까요?

jinwookh replied:
0과 100을 상수로 처리했습니다. 피드백을 제가 이해한 것이 맞나요?

jihan805 commented on src/main/java/lotto/domain/money/Prize.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.NoSuchElementException;
+
+public enum Prize {
+    FIRST(6, 2000000000),
```
Rank를 enum 이용해 구현 👍
보너스볼 여부를 판단하기 위해 bonus인지 아닌지를 판단하는 값을 추가해보면 어떨까요?

jinwookh replied:
bonus 수가 있는지 알려주는 boolean 변수를 아래와 같이 넣으면 될까요?
`FIRST(6, 2000000000, false),
    SECOND(5, 30000000, true),
    THIRD(5, 1500000, false),
    FOURTH(4, 50000, false),
    FIFTH(3, 5000, false),
    MISS(0, 0, false);`
제가 생각하기로, 위와 같이 bonus number가 있는지 알려주는 boolean값을 인자로 넣으면 2등이 bonus number를 가지고 있다는 사실을 표현하는데 도움이 될 것 같습니다.
다만, 1등, 4등, 꽝은 bonus number 유무와 관계없음에도 불구하고 코드에서 bonus number를 가지고 있지 않는 것처럼 표현됩니다. 
단지 표현상의 편의를 위해 bonus number 유무를 알리는 값을 추가하면 새로운 문제가 생길 것 같은데, Summer 리뷰어님은 어떻게 생각하시나요? bonus number 유무를 알리는 값을 통해 얻을 수 있는 장점이 또 있을까요?

jihan805 replied:
bonus 공의 유무를 알린다고 생각하기보다, bonus볼이 있는지에 대해 확인해보라는 의미로 이해해보면 어떨까요?
2등인 경우에만 보너스볼을 확인하면 될 것 같아요!
표현상의 편의를 위해 사용하기 보단, SECOND와 같이 특수한 경우에 대해 별도로 예외처리를 하지 않도록 값을 추가해보면 어떨까해서 피드백을 남겨보았습니다.
물론 로또의 경우 이미, 룰이 정해져있어 해당 되지 않을 것 같지만, 예를 들어  bonus볼을 확인해야하는 새로운 등수가 생겼을 경우 또 if () 와 같은 조건문을 추가하지 않고 구현할 수 있을 것 같아요!

jinwookh replied:
앗 bonus볼을 확인해야하는 새로운 등수를 구현해야 할 경우 bonus 유무 변수가 활용될 수 있겠네요! 답변 감사합니다.

jihan805 commented on src/test/java/lotto/ListTest.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public class ListTest {
+    @Test
```
꼭 필요한 테스트일까요?

jinwookh replied:
앗 미션과 관계없는 테스트를 넣었네요. 삭제했습니다.

jihan805 commented on src/test/java/lotto/LottoNumberTest.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @Test
+    void 숫자_범위_넘을때_테스트() {
```
경계값을 이용한 테스트 구현 👍 
1보다 작은 경우 대한 테스트도 구현해보면 어떨까요?

jinwookh replied:
추가했습니다!

jihan805 commented on src/test/java/lotto/LottosTest.java:
```diff
@@ -0,0 +1,49 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.offset;
+
+public class LottosTest {
```
전체적인 테스트 작성 💯 
LottoResult에서 작성해야 할 테스트들도 있는 것 같아요!
분리해보면 어떨까요?

jinwookh replied:
LottosTest 테스트 내용이 수익률에 관계된 테스트라서 LottoResult에 대한 테스트라고 해도 좋을 것 같아요. 파일명을 수정했습니다!

jihan805 commented on src/test/java/lotto/MoneyTest.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Test
+    void 천원단위_아닐때() {
```
해당 테스트는, 음수인 경우가 조금 더 적합할 것 같아요!
천원 단위가 아닌 경우에 대한 테스트는 별도로 작성하는게 어떨까요?

jinwookh replied:
넵!

jihan805 commented on src/test/java/lotto/WinningLottoTest.java:
```diff
@@ -0,0 +1,87 @@
+        assertThrows(IllegalNumberCombinationException.class, () -> {
+            WinningLotto.create(Arrays.asList(1, 2, 3, 4, 5, 6), 6);
+        });
+    }
```
불필요한 공백 제거!

jinwookh replied:
수정했습니다

jihan805 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private static void printLotto(Lotto lotto) {
+        System.out.println(lotto.toString());
```
toString을 오버라이드 했어요!
 System.out.println(lotto.toString()) -> System.out.println(lotto)와 같이 출력해볼까요?

jinwookh replied:
넵

jihan805 reviewed (changes requested):
안녕하세요 올레님, 
전체적으로 잘 구현하셨어요 👍 
몇가지 피드백 남겼으니 확인 부탁드려요 :)

jinwookh committed:
convert for to foreach at convertToLottoNumbers method

jinwookh committed:
change variable name lotto -> winningLotto at WinningLotto class

jinwookh committed:
make 0 and 100 constant

jinwookh committed:
removed ListTest class file

jinwookh committed:
change filename Lottos to LottoResult

jinwookh committed:
add bonus number info at Prize class

jinwookh committed:
add test case(when money does not devide into 1000)

jinwookh committed:
removed blank line

jinwookh committed:
remove unneccessary toString method inside of print method

jinwookh committed:
add message parameter into exception

jinwookh committed:
add exception package at each domain package

jinwookh committed:
rename getManualLottos to askManualLottos

jihan805 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,52 @@
+    public static List<LottoNumber> convertToLottoNumbers(List<Integer> numbers) {
+        List<LottoNumber> returnList = new ArrayList<>();
+        for (int number : numbers
+             ) {
```
)부분을 띄어쓴 특별한 이유가 있을까요?

jinwookh replied:
foreach문을 단축키로 쓰면 저렇게 나오는데, 이후 수정하지 않았습니다. 수정했습니다!

jihan805 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Lotto {
+    public static final int NUMBER_LENGTH = 6;
+    public static final int PRICE = 1000;
```
금액의 단위를 표현할때 1_000원과 같이 표현하면 훨씬 읽기 쉽지 않을까요?
Prize에도 적용해보면 좋을 것 같아요!

jihan805 replied:
위의 피드백에 대한 반영이 안된것 같아요 ㅠ
다음 단계 진행할때 같이 반영해주세요 :)

jinwookh replied:
깜빡 놓쳤습니다. 반영했습니다!

jihan805 reviewed (changes requested):
안녕하세요 올레님!
이미 구현을 잘하셔서인지 피드백 반영도 💯 
코멘트와 한가지 피드백 남겼으니 확인 부탁드릴게요 :)

jinwookh committed:
use automatic clean code refactoring

jinwookh committed:
use interface to implement automatic lotto generator and manual lotto generator

At first, LottoMachine class had whole logic of generating lotto logic.
However, we splitted generating logic into auto and manual logic, and made 2 additional classes for splitted logic.

jinwookh committed:
refactor by convention

jihan805 commented on src/main/java/lotto/domain/lotto/generator/AutomaticLottoGenerator.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+
+    private List<Integer> generateRandomLottoNumbers() {
```
불필요한 공백 제거!
클래스 전체적으로 불필요한 공백을 제거하면 어떨까요?

jihan805 reviewed (approved):
안녕하세요 올레님! 
피드백 반영 💯👍 머지할게요~!
다음 단계 진행하면서 반영하면 좋을 것 같은 간단한 피드백 남겼으니 확인 부탁드려요 :)

jinwookh commented:
친절한 리뷰 감사합니다!
